sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -364,48 +366,162 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+              df.queryExecution.analyzed.resolve(Seq(col), resolver).getOrElse {
+                val inputColumns = df.schema.map(_.name).mkString(", ")
+                throw new AnalysisException(
+                  s"cannot resolve '$col' given input columns: [$inputColumns]")
+              }
+            }
+
+            // Check if the specified partition columns match the existing table.
+            val specifiedPartCols = CatalogUtils.normalizePartCols(
+              tableName, tableCols, partitioningColumns.getOrElse(Nil), resolver)
+            if (specifiedPartCols != existingTable.partitionColumnNames) {
+              val existingPartCols = existingTable.partitionColumnNames.mkString(", ")
+              throw new AnalysisException(
+                s"""
+                   |Specified partitioning does not match that of the existing table $tableName.
+                   |Specified partition columns: [${specifiedPartCols.mkString(", ")}]
+                   |Existing partition columns: [$existingPartCols]
+                """.stripMargin)
+            }
+
+            // Check if the specified bucketing match the existing table.
+            val specifiedBucketSpec = getBucketSpec.map { bucketSpec =>
+              CatalogUtils.normalizeBucketSpec(tableName, tableCols, bucketSpec, resolver)
+            }
+            if (specifiedBucketSpec != existingTable.bucketSpec) {
+              val specifiedBucketString =
+                specifiedBucketSpec.map(_.toString).getOrElse("not bucketed")
+              val existingBucketString =
+                existingTable.bucketSpec.map(_.toString).getOrElse("not bucketed")
+              throw new AnalysisException(
+                s"""
+                   |Specified bucketing does not match that of the existing table $tableName.
+                   |Specified bucketing: $specifiedBucketString
+                   |Existing bucketing: $existingBucketString
+                """.stripMargin)
+            }
+
+            // Reorder the columns of the given dataframe to match the existing table.
+            val adjustedDataFrame = df.select(adjustedColumns.map(new Column(_)): _*)
+
+            if (classOf[CreatableRelationProvider].isAssignableFrom(existingProvider)) {
+              // For data source that implementing `CreatableRelationProvider`, we should call its
+              // `createRelation` to do the data appending. `CreatableRelationProvider` may not
+              // provide an `InsertableRelation`, so we can't use `DataFrameWriter.insertInto` here.
+              existingProvider.newInstance().asInstanceOf[CreatableRelationProvider].createRelation(
+                sqlContext = df.sparkSession.sqlContext,
+                mode = mode,
+                parameters = new CaseInsensitiveMap(extraOptions.toMap),
+                data = adjustedDataFrame)
+            } else {
+              adjustedDataFrame.write.insertInto(tableIdentWithDB)
```

Comment:
Currently only 2 data sources accept `saveAsTable` with append mode: `CreatableRelationProvider` and `FileFormat`. For `CreatableRelationProvider`, we always go to `createRelation`, for `FileFormat`, we go to `InsertIntoHadoopFsRelation`, which is same with `InsertIntoTable`. That's why I add the if-else here.